src/main/kotlin/org/wycliffeassociates/otter/jvm/persistence/repositories/WorkbookRepository.kt

```diff
@@ -0,0 +1,213 @@
+package org.wycliffeassociates.otter.jvm.persistence.repositories
+
+import com.jakewharton.rxrelay2.BehaviorRelay
+import com.jakewharton.rxrelay2.ReplayRelay
+import io.reactivex.Observable
+import io.reactivex.Single
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.rxkotlin.plusAssign
+import org.wycliffeassociates.otter.common.data.model.Collection
+import org.wycliffeassociates.otter.common.data.model.Content
+import org.wycliffeassociates.otter.common.data.model.MimeType
+import org.wycliffeassociates.otter.common.data.workbook.*
+import org.wycliffeassociates.otter.common.persistence.repositories.IWorkbookRepository
+import java.util.*
+
+class WorkbookRepository(
+    private val collectionRepository: CollectionRepository,
+    private val contentRepository: ContentRepository,
+    private val resourceRepository: ResourceRepository,
+    private val takeRepository: TakeRepository
+) : IWorkbookRepository {
+    /** Disposers for Relays in the current workbook. */
+    private val connections = CompositeDisposable()
+
+    /** Map to recover model.Take objects from workbook.Take objects. */
+    private val takeMap = WeakHashMap<Take, org.wycliffeassociates.otter.common.data.model.Take>()
```

Comment:
I suppose this works because the only takes added to the relay are what would be inserted? So even though null ids typically mean the take is new and can be inserted into the database, it's fine if they're lost through having to re-map a lost weakhashmap Take because we no longer need/care about the ids of existing Takes; updates to specific fields happen through their respective relays, and inserting new takes happens when a new Take is pushed into the relay (regardless of the contents of the id field)

Reply:
I scoped this to a single function so its purpose is more clear.